Add lxd container type to cloud + local providers

tych0 opened the pull request:
Hello folks,

Here's a patchset that adds lxd container type support for all the cloud providers + local provider. Some notes:

* This depends on my two other PRs #4131 and #4176; I'll be happy to rebase it when those are merged.
* I've only tested this on GCE (and local), but since I didn't have to make any GCE specific changes, hopefully it'll work everywhere else. Any testing/thoughts about what needs to be done for other providers are much appreciated.
* the lxd-images invocation here will need to change soon, as we're trying to get rid of lxd-images (probably next week those changes will land, and we'll have a go api for importing images). I can send a patch for that when we land it.
* I haven't added any more tests for this.
* I haven't gotten rid of the lxc container type. I assume we'll need to keep this so that we can support upgrades at some point.
* I don't know anything about juju, so it's entirely possible I've screwed all of this up :)

Thoughts appreciated!

jameinel commented on cmd/jujud/agent/machine.go:
```diff
@@ -1923,10 +1929,8 @@ func (a *MachineAgent) uninstallAgent(agentConfig agent.Config) error {
-	if err != nil {
-		errors = append(errors, err)
-	} else if insideLXC {
+	insideLXC := container.RunningInContainer()
```
Probably we should switch to "insideContainer" as the variable name for clarity.

jameinel commented on container/factory/factory.go:
```diff
@@ -21,6 +22,8 @@ func NewContainerManager(forType instance.ContainerType, conf container.ManagerC
 	switch forType {
 	case instance.LXC:
 		return lxc.NewContainerManager(conf, imageURLGetter, looputil.NewLoopDeviceManager())
+	case instance.LXD:
```
I would think we need similar support for LoopDeviceManager from LXC containers. But I suppose that can be on a TODO.

jameinel commented on container/lxd/initialisation.go:
```diff
@@ -0,0 +1,75 @@
+)
+
+var requiredPackages = []string{
+	"lxd",
```
Don't we also need things like "lxd-tools" to get lxd-images import?

tych0 replied:
lxd-images is in `lxd` proper, so I don't think so. Calling out to that will go away soon as well, and we'll just use the API to import images as soon as the LXD simplestreams code lands.

tych0 replied:
see, http://packages.ubuntu.com/xenial/amd64/lxd/filelist

jameinel commented on container/lxd/initialisation.go:
```diff
@@ -0,0 +1,75 @@
+
+	for _, pack := range requiredPackages {
+		pkg := pack
+		if config.SeriesRequiresCloudArchiveTools(series) &&
```
I'm not sure if we need this one, but I believe we *do* need to add something that knows to install "lxd/trusty-backports" on Trusty machines. IIRC CloudArchive was only needed for Precise? (And I'm guessing LXD isn't going to be available there.) I suppose you could still use a Trusty host with a Precise container?

tych0 replied:
I thought cloud archive existed for trusty too (and will probably exist for xenial when it comes time to push updated versions), but I'm not an expert here. I'll see about adding some code to do the backports thing, though, thanks.

jameinel commented on tools/lxdclient/client_image.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+func (i imageClient) EnsureImageExists(series string) error {
+	name := i.ImageNameForSeries(series)
```
do we have to worry about architecture? Or that is naturally determined just by the local host architecture?

tych0 replied:
Right, although it is possible to run i386 images on amd64 hosts, the architectures in general match. Of course we can support the i386 bit if that is necessary, but I suppose it isn't?

jameinel commented on tools/lxdclient/client.go:
```diff
@@ -52,11 +56,7 @@ var lxdLoadConfig = lxd.LoadConfig
-	defer updateLXDVars(origDirname)
-
-	cfg, err := lxdLoadConfig()
+	cfg, err := lxdLoadConfig(path.Join(configDir, "config.yml"))
```
Were we going to try and change this to be config strings that we pass in, rather than maintaining a somewhat arbitrary config directory?

tych0 replied:
Sure, but I don't know where/how to store it in juju's database, so I just left it as-is.

jameinel replied:
So we do have the issue that we don't generally let Providers (and thus Broker/container.Managers) have direct DB access. So we'd need to be passing this information in via just extra parameters to Init or Config.
What are the things we need, certificates and the config that says to use "local"?
I know we have support for some of those things at the Provider level, to date we haven't needed them for Brokers. Will need some thought as you need a custom cert for every machine to every LXD, right? How do we validate the LXD's certificate, or do we just accept it as long as it is valid?

tych0 replied:
On Tue, Jan 26, 2016 at 08:27:30PM -0800, John Arbash Meinel wrote:
> > @@ -52,11 +56,7 @@ var lxdLoadConfig = lxd.LoadConfig
> >  func newRawClient(remote, configDir string) (*lxd.Client, error) {
> >  	logger.Debugf("loading LXD client config from %q", configDir)
> >  
> > -	// This will go away once LoadConfig takes a dirname argument.
> > -	origDirname := updateLXDVars(configDir)
> > -	defer updateLXDVars(origDirname)
> > -
> > -	cfg, err := lxdLoadConfig()
> > +	cfg, err := lxdLoadConfig(path.Join(configDir, "config.yml"))
> 
> So we do have the issue that we don't generally let Providers (and thus Broker/container.Managers) have direct DB access. So we'd need to be passing this information in via just extra parameters to Init or Config.
> What are the things we need, certificates and the config that says to use "local"?

We don't need any certificates if you only want to talk to the LXD
provider on the local host, just enough perms to write to
/var/lib/lxd/unix.socket

> I know we have support for some of those things at the Provider level, to date we haven't needed them for Brokers. Will need some thought as you need a custom cert for every machine to every LXD, right? How do we validate the LXD's certificate, or do we just accept it as long as it is valid?

If you're only talking to the local lxd, since it's the only thing
that can listen on /var/lib/lxd/unix.socket, so there is no
authentication required. Things get a little more complicated when you
want to talk to remotes, but as far as I can tell we don't need to do
that (at least right now with the LXD provider/container type).

jameinel commented on provider/local/prereqs.go:
```diff
@@ -68,6 +69,8 @@ var VerifyPrerequisites = func(containerType instance.ContainerType) error {
 	case instance.LXC:
 		return verifyLxc()
+	case instance.LXD:
+		return nil
```
shouldn't this be the verifyLxd function you just implemented?

jameinel commented:
I think there are still a few steps that we'll need to add to bring it up to par with LXC. Things like creating template images and caching them in global state so that the time to bring up a large deployment with containers doesn't regress.

I'm also concerned that there aren't any new tests, even though there is a fair amount of new code being added. While I have some faith that Tycho is writing decent code, I'm concerned about future maintenance for people that don't understand this code as well, and may not know what the assumptions are.

tych0 commented:
On Tue, Jan 26, 2016 at 03:56:44AM -0800, John Arbash Meinel wrote:
> I think there are still a few steps that we'll need to add to bring it up to par with LXC. Things like creating template images and caching them in global state so that the time to bring up a large deployment with containers doesn't regress.

Yep, definitely.

> I'm also concerned that there aren't any new tests, even though there is a fair amount of new code being added. While I have some faith that Tycho is writing decent code, I'm concerned about future maintenance for people that don't understand this code as well, and may not know what the assumptions are.

Yep, I can add some tests for this (similar to what the LXC tests do).
Mostly I just wanted to get this in front of someone who knows what
needs to happen to review it. I'm traveling tomorrow, so hopefully
I'll have some time then to write/push some new tests. (I'll also be
on a european timezone at that point, so perhaps we can coordinate
more easily.)

tych0 commented:
Ok, I believe I've addressed all the review comments modulo the test bits which I'll try to address tomorrow (got some other kernel goop to debug today). I've also rebased things since there were patches that landed in master that conflicted.

jameinel commented:
I did just try to bootstrap this branch on EC2 with a Trusty target, and it failed because it couldn't install just "LXD". The worker that failed kept dying and trying again.
I then manually tried to "apt-get install lxd/trusty-backports" but that fails because it can't satisfy the rest of lxd's dependencies. I needed to do:
sudo apt-get install lxd/trusty-backports lxc/trusty-backports liblxc1/trusty-backports python3-lxc/trusty-backports

So we'll have to think carefully about how we enable trusty, given the need for more than just the obvious package.

tych0 commented:
On Wed, Jan 27, 2016 at 04:27:10AM -0800, John Arbash Meinel wrote:
> I did just try to bootstrap this branch on EC2 with a Trusty target, and it failed because it couldn't install just "LXD". The worker that failed kept dying and trying again.
> I then manually tried to "apt-get install lxd/trusty-backports" but that fails because it can't satisfy the rest of lxd's dependencies. I needed to do:
> sudo apt-get install lxd/trusty-backports lxc/trusty-backports liblxc1/trusty-backports python3-lxc/trusty-backports
> 
> So we'll have to think carefully about how we enable trusty, given the need for more than just the obvious package.

Right, this is a known bug (see the commit message), which apparently
hasn't been fixed yet :). I'll ping stgraber or someone about trying
to get an apt fix uploaded.

> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/juju/juju/pull/4191#issuecomment-175592174

tych0 commented:
On Wed, Jan 27, 2016 at 04:27:10AM -0800, John Arbash Meinel wrote:
> I did just try to bootstrap this branch on EC2 with a Trusty target, and it failed because it couldn't install just "LXD". The worker that failed kept dying and trying again.
> I then manually tried to "apt-get install lxd/trusty-backports" but that fails because it can't satisfy the rest of lxd's dependencies. I needed to do:
> sudo apt-get install lxd/trusty-backports lxc/trusty-backports liblxc1/trusty-backports python3-lxc/trusty-backports
> 
> So we'll have to think carefully about how we enable trusty, given the need for more than just the obvious package.

I guess that's a good point, though, that even with the bugfix we'll
need a newer LXC on trusty, which might break everything for users who
depend on a specific version of LXC in trusty.

ericsnowcurrently commented:
@tycho FYI, this repo has integration with reviewboard:

https://github.com/juju/juju/blob/master/CONTRIBUTING.md#code-review

However, that does not work until you have logged in to reviewboard the first time.  I'd recommend doing that so future PRs will automatically have review requests created.

tych0 commented:
Hi folks. I've pushed some sanity tests here, as well as fixed another bug. I think it would be good to land this so we can get people to start playing with it (and/or land #4131 so at least the 0.27 incompatibility is fixed). Thoughts?

cherylj commented:
@tych0 master is attempting to get a bless so we can ship alpha2.  Once we get a bless and ship alpha2, then we should merge this.

tych0 commented:
Ok. It's probably worth landing #4131 at least, then.

cherylj commented:
Could you guys drop this in a feature branch so it could get some CI testing before landing into master?

You will need to rebase as the api-command-rename branch has landed, and without those changes, nothing will pass CI.

tych0 commented:
On Wed, Feb 03, 2016 at 06:48:19AM -0800, Cheryl Jennings wrote:
> Could you guys drop this in a feature branch so it could get some CI testing before landing into master?

I don't know how to do that, exactly. Are there any docs on how you
guys do this?

> You will need to rebase as the api-command-rename branch has landed, and without those changes, nothing will pass CI.

Thanks for the heads up, I've rebased both this and #4131 in case you
want to land that one separately.

tych0 commented:
And rebased again :)

cherylj commented:
This PR is currently in feature branch lxd-container-type for CI testing and can be closed.  We will merge the feature branch into master when it has passed CI.
